### ISSUE TITLE
Build computational graph in bottom-up fashion

### ISSUE THREAD

**Issue by pcyin** [fchollet/keras#666]
Currently, Keras builds the computational graph in a `top-down` fashion. That is, starting from the final output tensor and recursively calling `get_output()` until reaching the input. This is fine for `sequential` models, since layers are stacked linearly and the computational flow is a straight line. However, for `graph` models, the computational flow could have branches, and this `top-down` fashion will incur redundancy because `self.get_output()` will "repeat" the downstream computations at the branching point. Consider the following example:

```
X = T.vector('X')
B = X * 3
Z1 = B + 1
Z2 = B + 2
Y = Z1 + Z2
theano.printing.pydotprint(Y)
```
The graph looks like [this](http://i.cs.hku.hk/~pcyin/graph1.png) (as one can imagine)

When implemented in Keras:
```
# helper layer
class LambdaLayer(Layer):
    def __init__(self, func):
        super(LambdaLayer, self).__init__()
        self.func = func

    def get_output(self, train=False):
        X = self.get_input(train)
        return self.func(X)

g = Graph()
g.add_input('X', ndim=1)

mul = LambdaLayer(lambda x: x * 3)
g.add_node(mul, name='mul', input='X')
add1 = LambdaLayer(lambda x: x + 1)
add2 = LambdaLayer(lambda x: x + 2)
add3 = LambdaLayer(lambda x: x['1'] + x['2'])

g.add_node(add1, name='add1', input='mul')
g.add_node(add2, name='add2', input='mul')
# I modified Keras to support multiple inputs, no Merge layer required here
g.add_node(add3, name='add3', inputs={'1': 'add1', '2': 'add2'}, create_output=True)

theano.printing.pydotprint(g.get_output())
```
The graph looks like [this](http://i.cs.hku.hk/~pcyin/graph2.png). `X * 3` repeats twice.

Of cause we can count on Theano's optimizer to reduce the redundancy, but when it comes to large graph models with complicated wiring. The optimization will take quite a long time (Actually I have a graph model with lots of branching and it takes 8 hours to compile). Does it seem better to build computational graph in bottom-up fashion?

**Comment by fchollet** [comment]
> Of cause we can count on Theano's optimizer to reduce the redundancy

Yes. In theory, all that matters is the graph topology, not how the graph was constructed.

> Actually I have a graph model with lots of branching and it takes 8 hours to compile

You are probably outside of what is reasonable to do in Theano. Don't worry though, Keras will move to a faster backend than Theano in the next 2-3 months. 

> Does it seem better to build computational graph in bottom-up fashion?

- How would it work concretely? (code examples)
- Is there any empirical evidence that the compilation or runtime would be faster?

**Comment by pranv** [comment]
I experienced exceptionally long times when compiling GoogLeNet type architectures which has 4^9 paths from the input to the cost.

I think bottom up fashion would be neat, if it can be done without compromising the simplicity of the current system.

**Comment by pranv** [comment]
https://github.com/Lasagne/Lasagne/issues/104 provides some direction here. 

Also, https://github.com/Lasagne/Lasagne/issues/97

**Comment by fchollet** [comment]
Interesting to note that the issues raised over at Lasagne arise from Theano-specific quirks, such as `RandomStreams` not behaving as a step-wise operation, and `scan` gradients not getting merged.

*In theory*, there should be no issue with building the computational graph top-bottom by calling `get_output` recursively. Intermediate redundant steps *should* be optimized away.

Before we change anything, we'll need to quantify how much of an issue this is with complex DAGs with Dropout / recurrent units.

**Comment by fchollet** [comment]
A simple preliminary test would be to test compiling / running a redundant Keras DAG with *and* without Dropout, and see if there is any speedup at compile time or runtime when eliminating Dropout.

**Comment by benanne** [comment]
> Intermediate redundant steps should be optimized away.

The problem seems to be that, even if they are correctly optimized away, it could take a very long time. Theano's optimization phase is already a bottleneck, so giving it more work than needed seems like a bad idea (although it would of course be convenient). 

w.r.t. dropout masks and other randomness, I believe the behaviour of Theano (not merging the subgraphs) is correct. It has no reason to assume that you want to reuse the same samples in separately generated sub-expressions, unless you explicitly tell it to.

I'll just say that for Lasagne this turned out to be a huge issue which made the implementation of some types of models impossible (highway networks for example). We switched from a recursive graph traversal approach to an iterative one (based on a topological ordering of the layers) for this reason.

**Comment by fchollet** [comment]
Thanks for the pointers. Iterating on a toposorted list of layers is a good solution, but I very much like the compositionality that recursiveness brings. 

A first naive approach would be sharing a model-wide cache between layers, something like:
```python
class Layer():
   def get_input():
      if self.prev_layer.ID in self.cache:
         return self.cache[self.prev_layer.ID]
      return self.prev_layer.get_output()

   def get_output():
      out = f(self.get_input())
      self.cache[self.ID] = out
      return out
```
The reference to the cache would be given to a layer when it gets added to a given container/model. A standalone subcontainer would have its own independent cache, which is fine since connections to such a subcontainer would only be via its final `get_output` method, and not to specific nodes inside the subcontainer.

Any possible issues with such an approach?

**Comment by benanne** [comment]
We considered memoization / caching as an alternative as well. This option was discussed somewhere in https://github.com/Lasagne/Lasagne/issues/104 iirc, although I appreciate it's a pretty long thread :) Maybe grep for some keywords.

**Comment by pcyin** [comment]
Thanks for keeping the discussion!

> The optimization will take quite a long time (Actually I have a graph model with lots of branching and it takes 8 hours to compile)

As mentioned in my previous post, I have a Keras graph model with many branching, which takes hours to compile. Now I rewrote the whole model based on Theano and build the computational graph in bottom-up way (just like the first code snippet). The new model just takes a few seconds to compile.

I am not sure whether my case a special one. It's possible that my graph just hits some hidden problems in Theano, which may cause the long compilation issue. 

> Iterating on a toposorted list of layers is a good solution

I agree :)

**Comment by nouiz** [comment]
I think part of the slowness is not just the duplicate computation before the grad, but that combined with the grad being computed on a graph with duplicate computation.

In blocks, this can happen with their replacement mechanism. One good solution to this would be to run the merge optimizer on the graph before calling grad.

Are you interrested to such an implementation?

**Comment by fchollet** [comment]
> One good solution to this would be to run the merge optimizer on the graph before calling grad. Are you interrested to such an implementation?

Yes, that sounds interesting! Also, what are your thoughts on layer output caching vs. flattening the network before building the Theano expression?

**Comment by nouiz** [comment]
I don't know enough your code to give good advice except this:

I think we should try to prevent duplicate code. In the pseudo-code you
gave you ask each layer to do code to cache this. I don't like that idea.
You could wrap the cache code in a decorator. Or you could have the code
that call the method do the caching itself?

Other then that, I think that going over the topological order of the
layers is simpler then the cache, and I prefer simpler solution when this
don't have bad consequence otherwise. But I didn't code it nor know your
code. So maybe this is more complication.

**Comment by nouiz** [comment]
Just to recap, I think that those 3 approchs would completly fix this problem:

-  topo order of the layers
- cache
- [use the merge optimizer before calling graph](https://github.com/Theano/Theano/issues/3393)

There is a ticket about the last. We will do it as we need it for blocks, but I don't know when it will get done.

Also, 1 and 2 will also speed up building the graph and are more safe, in the sense that we know that they will work for sure, but 3 we think it will work. But as it wasn't done, there could be  details that cause unforeseen consequence. But the only one I can think of is that it will be slower then option 1 and 2. Not sure how important that is.

**Comment by nouiz** [comment]
Please, follow that ticket to know about what happen to it.

**Comment by nouiz** [comment]
I'll update the merge ticket Monday about this. But doing that before grad
require that you don't use consider constant or known grad or that the wrt
variables won't me a variation that is merged(inputs and shared var are
fine). If any of those condition are true then the implementation get much
more complicated.
Le 25 sept. 2015 14:20, "Frédéric Bastien" <frederic.bastien@gmail.com> a
écrit :

> I don't know enough your code to give good advice except this:
>
> I think we should try to prevent duplicate code. In the pseudo-code you
> gave you ask each layer to do code to cache this. I don't like that idea.
> You could wrap the cache code in a decorator. Or you could have the code
> that call the method do the caching itself?
>
> Other then that, I think that going over the topological order of the
> layers is simpler then the cache, and I prefer simpler solution when this
> don't have bad consequence otherwise. But I didn't code it nor know your
> code. So maybe this is more complication.
>
>
>
> On Fri, Sep 25, 2015 at 2:03 PM, François Chollet <

**Comment by nouiz** [comment]
I think you are better to implement the building of the graph bottom-up in Keras. The pre-merge before the grad idea won't be done shortly and we will probably fix our related problems otherwise.

**Comment by fchollet** [comment]
We introduced output caching, thereby fixing this issue. Also yields faster compilation times, as expected.

Thank you for the valuable issue report.

**Comment by nouiz** [comment]
Out of curiosity, which PR included that? I'm just wanting to have an idea
how it was implemented.

**Comment by fchollet** [comment]
@nouiz the implementation is really simple.

1. Every container maintains a `self.layer_cache` dict
2. Every layer that gets added to the container gets a reference to the parent container's `layer_cache`
3. `layer.get_input(train)` is modified to the following:

```python
    def get_input(self, train=False):
        if hasattr(self, 'previous'):
            # to avoid redundant computations,
            # layer outputs are cached when possible.
            if hasattr(self, 'layer_cache'):
                previous_layer_id = '%s_%s' % (id(self.previous), train)
                if previous_layer_id in self.layer_cache:
                    return self.layer_cache[previous_layer_id]
            previous_output = self.previous.get_output(train=train)
            if hasattr(self, 'layer_cache'):
                previous_layer_id = '%s_%s' % (id(self.previous), train)
                self.layer_cache[previous_layer_id] = previous_output
            return previous_output
```

This implementation assumes thread-safeness. Hopefully that's a reasonable assumption?

**Comment by nouiz** [comment]
I don't recall anybody tring to build the graph in parallel. Also, I don't
think Theano allow parallel compilation of Theano functions.

So this seem a resonable assumption.

On Fri, Dec 11, 2015 at 11:07 AM, François Chollet <notifications@github.com

**Comment by around1991** [comment]
Hi @fchollet, I think there's an issue with how outputs get cached - particularly with the hash value used to retrieve cached outputs. There's a long discussion (with me rambling a lot) in #1254, but the long and short of it is that:

If a layer container has multiple dynamically set inputs, then as outputs are hashed based only on the ID of the previous node, every layer but the first layer in the container will retrieve the wrong output if the input is changed. This causes model compilation errors if you build a Graph model with two inputs and a shared stack of layers.

There's some discussion of a solution in #1254 as well.

Hope this is useful.

**Comment by neggert** [comment]
What do you think about caching input_shape as well? I'm trying to build an inception_v3 model and it takes quite a long time to even construct the model (before compiling it) because in later layers, calculating input/output shape recurses through the entire structure and takes forever.
